src/slurm_plugin/instance_manager.py

```diff
@@ -619,3 +566,576 @@ def _create_request_for_nodes(table_name, node_names):
+                self.unused_launched_instances |= instances_launched
+                self._update_failed_nodes(set(job.nodes_resume), "LimitedInstanceCapacity")
+            else:
+                # No instances launched at all, e.g. CreateFleet API returns no EC2 instances
+                self._update_failed_nodes(set(job.nodes_resume), "InsufficientInstanceCapacity")
+        else:
+            # Not implemented, never goes here
+            logger.error("Best effort job level scaling not implemented")
+
+    def _launch_instances_for_job(
+        self, job: SlurmResumeJob, nodes_to_launch: Dict[str, any], launch_batch_size: int, all_or_nothing_batch: bool
+    ):
+        instances_launched = collections.defaultdict(lambda: collections.defaultdict(list))
+        for queue, compute_resources in nodes_to_launch.items():
+            for compute_resource, slurm_node_list in compute_resources.items():
+                slurm_node_list = self._resize_slurm_node_list(
+                    queue=queue,
+                    compute_resource=compute_resource,
+                    instances_launched=instances_launched,
+                    slurm_node_list=slurm_node_list,
+                )
+
+                if slurm_node_list:
+                    logger.info("Launching instances for nodes %s", print_with_count(slurm_node_list))
+                    # Set the number of retries to be the max between the globally configured one and 3.
+                    # This is done to try to avoid launch instances API throttling
+                    # without changing the configured retries for all API calls.
+                    configured_retry = (
+                        self._boto3_config.retries.get("max_attempts", 0) if self._boto3_config.retries else 0
+                    )
+                    boto3_config = self._boto3_config.merge(
+                        Config(retries={"max_attempts": max([configured_retry, 3]), "mode": "standard"})
+                    )
+                    # Each compute resource can be configured to use create_fleet or run_instances
+                    fleet_manager = FleetManagerFactory.get_manager(
+                        cluster_name=self._cluster_name,
+                        region=self._region,
+                        boto3_config=boto3_config,
+                        fleet_config=self._fleet_config,
+                        queue=queue,
+                        compute_resource=compute_resource,
+                        all_or_nothing=all_or_nothing_batch,
+                        run_instances_overrides=self._run_instances_overrides,
+                        create_fleet_overrides=self._create_fleet_overrides,
+                    )
+
+                    for batch_nodes in grouper(slurm_node_list, launch_batch_size):
+                        try:
+                            launched_ec2_instances = fleet_manager.launch_ec2_instances(
+                                len(batch_nodes), job_id=job.job_id
```

Comment:
I couldn't find a good way, because fleet_manager instantiate it's own logger. I could have returned the fleet_manager logger object when calling the get_manager, but I didn't like the approach.

For the case of node list scaling, the job_id will be default to None